### ISSUE TITLE
"When specifying '_scheme', '_external' must be True."

### ISSUE THREAD

**Issue by debben** [briantist/galactory#122]
When using galactory with `preferred-url-scheme` option, I was consistently getting the following exception after upgrading to 0.11.1:

```
raise ValueError("When specifying '_scheme', '_external' must be True.")
```
coming from invocations of `url_for` in galactory/api/v3/collections.py. I noticed in v3 these are hardcoded `_external=False` but in the v2 implementation of collections.py all of the `_external` parameters of `url_for` are set to True. Setting these to True in a local development environment fixed the behavior. I making this issue to bring awareness and make sure I'm not just misconfiguring galactory when using v3.

**Comment by briantist** [comment]
Hi @debben ! Thanks for raising this, I don't usually use this option and this project is embarrassingly short on good tests right now so I didn't catch this.

`_external` was changed from `True` to `False` between v2 and v3 because the v2 API returns absolute URLs and v3 returns relative. Here's a demonstration of the difference:
- https://old-galaxy.ansible.com/api/v2/collections/community/general/
- https://galaxy.ansible.com/api/v3/plugin/ansible/content/published/collections/index/community/general/

The proper fix for this is probably not to set `_external` to `True` in v3, but rather to stop passing `_scheme` at all since v3 uses relative URLs, and specifying a scheme is not needed for relative URLs.

---

I would like to learn more about your use of the feature though. I think in my experience it's not usually needed, so if I could understand it better that'd be really helpful.

**Comment by debben** [comment]
Thanks @briantist. I had set the parameter in a local environment to http for localhost testing. In a production environment I also had it set (though to https). Just removing the option altogether from the live configuration was sufficient to fix the error.

Now I'm facing a different issue though in that live environment where api requests to collections coming from my galaxy client work fine (https) but url the galaxy attempts to download the tarball from is http.

**Comment by briantist** [comment]
The download URL is the one generated URL in v3 that's still absolute, so it makes sense that if you needed the option before, you would need it again.

Can you tell me a bit more about your production environment? All of your (non-secret) configuration options, and which WSGI server you're using (or if you're using the built-in werkzeug)?

If the issue is that you're using a reverse proxy / load balancer, then see: 
- https://github.com/briantist/galactory/pull/48

---

In any case, I will be fixing the bug you reported here which should allow you to continue using `--preferred-url-scheme` as intended, I expect to have a fix maybe tomorrow? If you'd like you could also put up a PR, I believe simply removing the `_scheme` parameter from any `url_for` calls that use `_external=False` should fix it.

Any `url_for` calls that use `_external=True` should continue to set `_scheme`.

**Comment by debben** [comment]
For now I'm just hosting the app in a container in a kubernetes cluster. There's an ingress handling ssl termination, but I've only got it setup to listen on 443. I'm going to see if I can setup an http listener on that ingress to 302 everything to https endpoint. I'm also going to attempt to append the x-forwarded-proto header you mentioned and see if that will get download urls coming back as https.

**Comment by briantist** [comment]
I'm not too familiar with k8s but there's a chance the ingress is already setting the header. If it is, or if you can set it to, then you will need to change the container invocation so that it uses the proxy_fix method. I don't recall if it can work with the built-in werkzeug webserver, but I can say at least anecdotally that I use it with gunicorn and it's really simple.

I still haven't gotten around to starting #57 , but here's a sample Dockerfile for building your own container that uses gunicorn:

```dockerfile
ARG galactory_tag=latest

FROM ghcr.io/briantist/galactory:${galactory_tag}

ENV PYTHONUNBUFFERED=TRUE

RUN pip install 'gunicorn[gevent]>=20,<=21'

COPY gunicorn.conf.py /etc/gunicorn.d/
COPY galactory.conf /root/.galactory/

ENTRYPOINT [ "gunicorn", "--config=/etc/gunicorn.d/gunicorn.conf.py", "galactory:create_configured_app(proxy_fix=True, x_for=1, x_proto=1)" ]
```

And here's a sample `gunicorn.conf.py`
```python
bind = "0.0.0.0:5555"
workers = 4
threads = 2
worker_class = "gevent"
capture_output = True
accesslog = "-"
errorlog = "-"
```

`galactory.conf` is a standard config file for galactory, you could leave that out and set options on the command line with the gunicorn implementation, or with environment variables instead, but I think the config file method is better.

Just remember that numbers in the `x_` parameters need to match the number of hops (proxies), that might be more than 1 if there are multiple proxies or load balancers involved.

**Comment by debben** [comment]
Opened #123 for the change you mentioned. The upgrading of http via 302 isn't really an option for me. I've tried the x-forwarded-proto but also my familiarity with k8s (at least this particular ingress) is lacking so I'm not sure if the app is getting the header. I'll continue to investigate and might give this gunicorn approach a try.

**Comment by briantist** [comment]
Thank you for the PR! I will review it as soon as I can.

Even if the header is being set, it won't be used unless you use the proxy_fix method. If you want to confirm which headers are sent to galactory, you can set `--log-level DEBUG` along with `--log-headers` and it will output them to the console. Be aware that this will leak credentials if auth is sent from the galaxy client so mind any logging you have.

**Comment by briantist** [comment]
Fixed by #123

**Comment by briantist** [comment]
Thanks for your help with this @debben , this fix has been released: https://github.com/briantist/galactory/releases/tag/v0.11.2
